File: C1.UWP.FlexChart/VB/FinancialChartExplorer/Services/DataService.vb
Imports System.Runtime.Serialization
Imports System.Runtime.Serialization.Json
Imports System.Collections.Generic
Imports System.Reflection
Imports System.Linq

Public Class DataService
    Private _companies As New List(Of Company)()
    Private _cache As New Dictionary(Of String, List(Of Quote))()

    Private Sub New()
        _companies.Add(New Company() With {
            .Symbol = "box",
            .Name = "Box Inc"
        })
        _companies.Add(New Company() With {
            .Symbol = "fb",
            .Name = "Facebook"
        })
    End Sub

    Public Function GetCompanies() As List(Of Company)
        Return _companies
    End Function

    Public Function GetData(Of T)(symbol As String) As List(Of T)
        Dim path As String = String.Format("FinancialChartExplorer.{0}.json", symbol)
        Dim asm As Assembly = Me.[GetType]().GetTypeInfo().Assembly
        Dim stream As Stream = asm.GetManifestResourceStream(path)
        Dim ser As New DataContractJsonSerializer(GetType(T()))
        Dim data As T() = CType(ser.ReadObject(stream), T())
        Return data.ToList()
    End Function


    Public Function GetSymbolData(symbol As String, Optional count As Integer = 20) As List(Of Quote)
        If Not _cache.Keys.Contains(symbol) Then
            Dim dataList As List(Of Quote) = GetData(Of Quote)(symbol)
            _cache.Add(symbol, dataList)
        End If
        If IsWindowsPhoneDevice() Then
            Dim dataList As List(Of Quote） = CType(_cache(symbol), List(Of Quote))
            Return dataList.Take(count).ToList()
        End If
        Return _cache(symbol)
    End Function

    Function IsWindowsPhoneDevice() As Boolean
        If Windows.Foundation.Metadata.ApiInformation.IsTypePresent("Windows.Phone.UI.Input.HardwareButtons") Then
            Return True
        End If
        Return False
    End Function

    Public Function FindRenderedYRange(data As List(Of Quote), xmin As Double, xmax As Double) As Range
        Dim item As Quote
        Dim ymin As Double = Double.NaN
        Dim ymax As Double = Double.NaN

        Dim i As Integer = 0
        While i < data.Count
            item = data(i)
            If xmin > i OrElse i > xmax Then
                i += 1
                Continue While
            End If
            If Double.IsNaN(ymax) OrElse item.High > ymax Then
                ymax = item.High
            End If
            If Double.IsNaN(ymin) OrElse item.Low < ymin Then
                ymin = item.Low
            End If
            i += 1
        End While

        Return New Range() With {
        .Min = ymin,
        .Max = ymax
    }
    End Function

    Public Function FindApproxRange(min As Double, max As Double, percent As Double) As Range
        Dim range As Range = New Range()
        range.Max = max
        range.Min = (max - min) * (1 - percent) + min
        Return range
    End Function

    Shared _ds As DataService
    Public Shared Function GetService() As DataService
        If _ds Is Nothing Then
            _ds = New DataService()
        End If
        Return _ds
    End Function
End Class

<DataContract>
Public Class Quote
    <DataMember(Name:="date")>
    Public Property [Date]() As String

    <DataMember(Name:="high")>
    Public Property High() As Double

    <DataMember(Name:="low")>
    Public Property Low() As Double

    <DataMember(Name:="open")>
    Public Property Open() As Double

    <DataMember(Name:="close")>
    Public Property Close() As Double

    <DataMember(Name:="volume")>
    Public Property Volume() As Double

    Public ReadOnly Property [Date2] As DateTime
        Get
            Return System.DateTime.ParseExact([Date].ToString, "MM/dd/yy", System.Globalization.CultureInfo.InvariantCulture)
        End Get
    End Property

End Class

<DataContract>
Public Class Annotation
    <DataMember(Name:="title")>
    Public Property Title() As String

    <DataMember(Name:="description")>
    Public Property Description() As String

    <DataMember(Name:="publicationDate")>
    Public Property PublicationDate() As String

    <DataMember(Name:="dataIndex")>
    Public Property DataIndex() As Integer
End Class

Public Class Range
    Public Property Min() As Double
    Public Property Max() As Double
End Class

Public Class Company
    Public Property Symbol() As String
    Public Property Name() As String
End Class

File: C1.UWP.FlexChart/CS/FinancialChartExplorer/Services/DataService.cs
using System.Linq;
using System.Reflection;
using System.Collections.Generic;
using System.Runtime.Serialization.Json;
using System.Runtime.Serialization;

namespace FinancialChartExplorer
{
    public class DataService
    {
        List<Company> _companies = new List<Company>();
        Dictionary<string, List<Quote>> _cache = new Dictionary<string, List<Quote>>();

        private DataService()
        {
            _companies.Add(new Company() { Symbol = "box", Name = "Box Inc" });
            _companies.Add(new Company() { Symbol = "fb", Name = "Facebook" });
        }

        public List<Company> GetCompanies()
        {
            return _companies;
        }

        public List<T> GetData<T>(string symbol)
        {
            string path = string.Format("FinancialChartExplorer.Resources.{0}.json", symbol);
            var asm = this.GetType().GetTypeInfo().Assembly;
            var stream = asm.GetManifestResourceStream(path);
            var ser = new DataContractJsonSerializer(typeof(T[]));
            var data = (T[])ser.ReadObject(stream);

            return data.ToList();
        }

        public List<Quote> GetSymbolData(string symbol, int count = 20)
        {
            if (!_cache.Keys.Contains(symbol))
            {
                var dataList = GetData<Quote>(symbol);
                _cache.Add(symbol, dataList);
            }
            if (IsWindowsPhoneDevice())
            {
                var dataList = _cache[symbol] as List<Quote>;
                return dataList.Take(count).ToList();
            }
            return _cache[symbol];
        }

        public bool IsWindowsPhoneDevice()
        {
            if (Windows.Foundation.Metadata.ApiInformation.IsTypePresent("Windows.Phone.UI.Input.HardwareButtons"))
            {
                return true;
            }
            return false;
        }

        public Range FindRenderedYRange(List<Quote> data, double xmin, double xmax)
        {
            Quote item;
            double ymin = double.NaN;
            double ymax = double.NaN;

            for (int i = 0; i < data.Count; i++)
            {
                item = data[i];
                if (xmin > i || i > xmax)
                {
                    continue;
                }
                if (double.IsNaN(ymax) || item.High > ymax)
                {
                    ymax = item.High;
                }
                if (double.IsNaN(ymin) || item.Low < ymin)
                {
                    ymin = item.Low;
                }
            }

            return new Range() { Min = ymin, Max = ymax };
        }

        public Range FindApproxRange(double min, double max, double percent)
        {
            var range = new Range();
            range.Max = max;
            range.Min = (max - min) * (1 - percent) + min;
            return range;
        }

        static DataService _ds;
        public static DataService GetService()
        {
            if (_ds == null)
                _ds = new DataService();
            return _ds;
        }
    }

    [DataContract]
    public class Quote
    {
        [DataMember(Name = "date")]
        public string Date { get; set; }

        [DataMember(Name = "high")]
        public double High { get; set; }

        [DataMember(Name = "low")]
        public double Low { get; set; }

        [DataMember(Name = "open")]
        public double Open { get; set; }

        [DataMember(Name = "close")]
        public double Close { get; set; }

        [DataMember(Name = "volume")]
        public double Volume { get; set; }

        public System.DateTime date
        {
            get { return System.DateTime.ParseExact(Date.ToString(), "MM/dd/yy", System.Globalization.CultureInfo.InvariantCulture); }
        }
    }

    [DataContract]
    public class Annotation
    {
        [DataMember(Name = "title")]
        public string Title
        {
            get; set;
        }

        [DataMember(Name = "description")]
        public string Description
        {
            get; set;
        }

        [DataMember(Name = "publicationDate")]
        public string PublicationDate
        {
            get; set;
        }

        [DataMember(Name = "dataIndex")]
        public int DataIndex
        {
            get; set;
        }
    }

    public class Range
    {
        public double Min { get; set; }
        public double Max { get; set; }
    }

    public class Company
    {
        public string Symbol { get; set; }
        public string Name { get; set; }
    }
}
